addons/udes_stock/models/stock_move_line.py

```diff
@@ -366,10 +360,152 @@ def prepare(
                     mls |= new_ml
         return dict(res)
 
+    def _swap_tracked_items(
+        self, mls, quantity, product, location, is_serial, vals, res, lot_names
+    ):
+        """
+        This handles the tracked product swaps. A tracked product swap is when a different lot name, than what is on the move_lines,
+        has been passed to prepare(). The swapped in lot name should be in the same location and of the same product, else no swaps
+        will happen.This method will produce a dictionary of a dictionary containing
+        update information on different movelines so that the tracked products can be swapped. An adjusted list of
+        lot_names is also returned.
+        Returns: dict(dict), list()
+        """
+        StockProductionLot = self.env["stock.production.lot"]
+        StockMoveLine = self.env["stock.move.line"]
+        Quant = self.env["stock.quant"]
+
+        # Swapping the tracked products
+        mls = mls.filtered(lambda m: m.qty_done < m.product_uom_qty)
+        lot_names_on_picking = mls.lot_id.mapped("name")
+        lot_names_swapped_in = list(set(lot_names) - set(lot_names_on_picking))
+        # If there swapping tracked products are allowed and some lot_names have been swapped in
+        if mls.u_picking_type_id.u_allow_swapping_tracked_products:
+            mls_not_changing = mls.filtered(lambda mls: mls.lot_id.name in lot_names)
+
+            # Needed for the special case where we overpick on a moveline on the pick and
+            # want the overpicked quantity to be swapped in on the other movelines
+            for ml in mls_not_changing:
+                # Test that there are other movelines for the overpicking fulfillment
+                # ml.product_uom_qty < quantity
+                if (
+                    not is_serial
+                    and ml.product_uom_qty < quantity
+                    and not lot_names_swapped_in
+                    and len(mls) > 1
+                ):
+                    mls_not_changing = mls_not_changing - ml
+                    lot_names_swapped_in.append(ml.lot_id.name)
+
+            quantity_swapping = quantity - sum(mls_not_changing.mapped("product_uom_qty"))
+            mls_need_changing = mls - mls_not_changing
+            if quantity_swapping and lot_names_swapped_in:
+                picking_mls, new_ml = mls._find_move_lines(
+                    quantity_swapping, product, None, None, location
+                )
+
+                # Workout the lot_ids that were swapped in and store the ones that were swapped out
+                lot_ids_swapped_in = StockProductionLot.search(
+                    [("name", "in", lot_names_swapped_in)]
+                )
+                lot_ids_swapped_out = picking_mls.lot_id
+
+                # If any of the lot_ids_swapped_in belong to another move_line
+                swapped_lot_on_assigned_mls = StockMoveLine.search(
```

Comment:
I don't think we care of the order of the move lines, so probably better to order by id.